cpp/src/sampling/detail/gather_one_hop_edgelist_impl.cuh

```diff
@@ -48,400 +144,551 @@ struct return_edges_with_properties_e_op {
-  std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
-  std::optional<edge_property_view_t<edge_t, edge_t const*>> edge_id_view,
-  std::optional<edge_property_view_t<edge_t, edge_type_t const*>> edge_type_view,
-  cugraph::vertex_frontier_t<vertex_t, tag_t, multi_gpu, false> const& vertex_frontier,
-  std::optional<raft::host_span<uint8_t const>> gather_flags,
-  bool do_expensive_check)
-{
-  rmm::device_uvector<vertex_t> majors(0, handle.get_stream());
-  rmm::device_uvector<vertex_t> minors(0, handle.get_stream());
-  std::optional<rmm::device_uvector<edge_t>> edge_ids{std::nullopt};
-  std::optional<rmm::device_uvector<weight_t>> edge_weights{std::nullopt};
-  std::optional<rmm::device_uvector<edge_type_t>> edge_types{std::nullopt};
-  std::optional<rmm::device_uvector<label_t>> labels{std::nullopt};
-
-  using edge_value_t = std::conditional_t<
-    has_weight,
-    std::conditional_t<
-      has_edge_id,
-      std::conditional_t<has_edge_type,
-                         thrust::tuple<weight_t, edge_t, edge_type_t>,
-                         thrust::tuple<weight_t, edge_t>>,
-      std::conditional_t<has_edge_type, thrust::tuple<weight_t, edge_type_t>, weight_t>>,
-    std::conditional_t<
-      has_edge_id,
-      std::conditional_t<has_edge_type, thrust::tuple<edge_t, edge_type_t>, edge_t>,
-      std::conditional_t<has_edge_type, edge_type_t, cuda::std::nullopt_t>>>;
-
-  using edge_value_view_t = edge_property_view_type_t<edge_t, edge_value_t>;
-
-  edge_value_view_t edge_value_view{};
-  if constexpr (has_weight) {
-    if constexpr (has_edge_id) {
-      if constexpr (has_edge_type) {
-        edge_value_view = view_concat(*edge_weight_view, *edge_id_view, *edge_type_view);
-      } else {
-        edge_value_view = view_concat(*edge_weight_view, *edge_id_view);
-      }
+struct gather_one_hop_edgelist_functor_t {
+  raft::handle_t const& handle;
+  graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view;
+  key_bucket_t<vertex_t, tag_t, multi_gpu, false> const& key_list;
+  std::optional<raft::host_span<uint8_t const>> gather_flags;
+  bool do_expensive_check{false};
+
+  template <typename TupleType>
+  auto concatenate_views(TupleType edge_properties)
+  {
+    if constexpr (std::tuple_size_v<TupleType> == 0) {
+      return edge_property_view_type_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>,
+                                       cuda::std::nullopt_t>{};
+    } else if constexpr (std::tuple_size_v<TupleType> == 1) {
+      return std::get<0>(edge_properties);
     } else {
-      if constexpr (has_edge_type) {
-        edge_value_view = view_concat(*edge_weight_view, *edge_type_view);
-      } else {
-        edge_value_view = *edge_weight_view;
-      }
+      return view_concat(edge_properties);
     }
-  } else {
-    if constexpr (has_edge_id) {
-      if constexpr (has_edge_type) {
-        edge_value_view = view_concat(*edge_id_view, *edge_type_view);
+  }
```

Comment:
Should we replicate this concatentate_views() function in every functor? They look very similar and can't we avoid this replication?